lib/ansible/module_utils/remote_management/spacewalk.py

```diff
@@ -0,0 +1,329 @@
+        try:
+            self.client.channel.software.removeRepo(self.session, label)
+        except Exception as generic_exception:
+            self.module.fail_json(repository=label, action='Delete', msg='Error deleting repository: %s' % generic_exception)
+
+    def repository_comparison(self, label):
+        """ Compare playbook inputs with repository in spacewalk and return the differences
+        """
+        changes = {}
+        key_map = {'label': 'label',
+                   'repourl': 'sourceUrl',
+                   'repotype': 'type',
+                   'sslCaCert': 'sslCaCert',
+                   'sslCliCert': 'sslCliCert',
+                   'sslCliKey': 'sslCliKey'
+                   }
+
+        if not self.repositories['spacewalk_repositories']:
+            self.get_all_repositories()
+
+        if label not in self.repositories['spacewalk_repositories']:
+            self.module.fail_json(msg="Error: no repository %s in repositories" % label)
+
+        repo = self.repositories['spacewalk_repositories'][label]
+
+        for keypb, keyrepository in key_map.items():
+            if keyrepository in repo:
+                pb_value = self.module.params[keypb]
+                rp_value = repo[keyrepository]
+
+                if pb_value != rp_value:
+                    changes[keyrepository] = {'original_value': rp_value, 'new_value': pb_value}
+
+        return changes
+
+    def set_repository_url(self, label, url):
+        try:
+            self.client.channel.software.updateRepoUrl(self.session, label, url)
+        except Exception as generic_exception:
+            self.module.fail_json(repository=label, action='Update', msg='Error setting url: %s' % generic_exception)
+
+    def set_repository_label(self, repoid, label):
+        try:
+            self.client.channel.software.updateRepoLabel(self.session, repoid, label)
+        except Exception as generic_exception:
+            self.module.fail_json(repository=label, action='Update', msg='Error setting label: %s' % generic_exception)
+
+    def set_repository_ssl(self, label, sslCaCert, sslCliCert, sslCliKey):
+        try:
+            self.client.channel.software.updateRepoUrl(self.session, label, sslCaCert, sslCliCert, sslCliKey)
+        except Exception as generic_exception:
+            self.module.fail_json(repository=label, action='Update', msg='Error setting ssl: %s' % generic_exception)
+
+    def update(self, label, changes):
+        if 'sourceUrl' in changes:
+            url = changes['sourceUrl']['new_value']
+            self.set_repository_url(label, url)
+            return 1
+        self.module.fail_json(repository=label, action='Update', msg='Error Change not implemented yet: %s' % changes)
+
+
+def connect_to_api(module, disconnect_atexit=True):
+    url = module.params['url']
+    login = module.params['login']
+    password = module.params['password']
+    validate_certs = module.params['validate_certs']
+
+    if not url:
```

Comment:
I leave inputs required because, as user, when I read docs I'm looking for input requirements. You have to pass it as input playbook or variable enviroment. but if you omit this the module will fails. In my mind I think best practice could be use inputs in playbook and use env var for cli using ansible -m. But this is just my opinion. Let me know what do you think about, I can remove requires

Reply:
But this code is never going to be hit if the argument_spec is set to required=True, unless someone set `url: ""`, but it's difficult to test for every bad input format - someone could equally pass `url: "not_a_protocol://thisisnonsense"`

Reply:
Basically @gundalow's point is that you don't need to repeat the work that Ansible is already doing enforcing required parameters

Reply:
Ok thank you. Checks removed